design/secure-credentials-store.md

```diff
@@ -21,51 +21,119 @@ Credentials include:
+place where a secret is retrieved should support a way to obtain it from the vault that needs to be
+consistent across the whole product.
+
+Generally, there might be more than one vault SPI provider available at the same time.
+For example, when a new vault is plugged in addition to an existing one, the values
+from the original vault are likely not be migrated immediately, and thus the references
+to the original vault should work.
+
+### Realm Keys
+
+Today there are SPIs both for realm keys as well as signing of tokens. Keys SPI enables storing keys in Java KeyStore as well as adding custom support for a vault. Signature SPI enables delegating token signature to external HSM devices.
+
+## Design
+
+### Vault Provider SPI
+
+It may be beneficial to allow per-type configuration. For example retrieving LDAP and SMTP credentials from Kubernetes secrets, while using HashiCorp Vault Transit Secrets for external tokens and not encrypting user secrets like otp secrets. That would allow more flexiblity as well as tuning security vs performance.
+
+Vault will manage access to vault providers and ensure proper vault is used to obtain the secret.
+
+For the consumer, the interface would be like this:
+
+```java
+public interface VaultProvider {
+
+    public static class Purpose {
+
+        public static final String REALM_KEY = "realm-key";
+        public static final String SMTP_SECRET = "smtp";
+        public static final String LDAP_SECRET = "ldap";
+        public static final String USER_CREDENTIAL = "broker-credential";
+        public static final String IDENTITY_BROKER_CREDENTIAL = "user-credential";
+        public static final String IDENTITY_BROKER_EXTERNAL_TOKEN = "broker-token";
+    }
+
+    /**
+     * Retrieves a secret from vault. The value returned by the first vault provider that resolves
+     * the secret successfully (even to {@code null}) is returned;  if no vault provider resolves it,
+     * this method results into an empty {@link java.util.Optional}.
+     *
+     * @param context Context that contains info about current request, to provide e.g. realm separation
+     * @param vaultSecretId Identifier of the secret. It corresponds to the value entered by user
+     *        in the respective configuration, which in turn is obtained from the vault when storing
+     *        the secret.
+     * @param purpose Purpose of the secret, set by keycloak according to the caller context.
+     * @return The secret or {@code null} if the secret was successfully resolved, or
+     *         an empty {@link java.util.Optional}
+     */
+    Optional<byte[]> obtainSecret(KeycloakContext context, String vaultSecretId, String purpose);
+}
+```
+
+Vault configuration has to contain mapping between purpose and the allowed provider IDs sorted by priority to be used for this purpose, e.g. `realm-key => HSM, vault-1`, `smtp => vault-1`, `default => vault-2, HSM` would mean that:
```

Comment:
I have some doubts how this approach will scale if we add more features to Keycloak. Note, that every time you add something new to Keycloak that requires storing some sort of credentials, you will need to revisit this mapping and add new parameters to it.

How about doing this differently. Imagine, that you're storing SMTP credentials and you'd like to store both, username and password in a vault. A Keycloak Admin could do it in the Email configuration screen by setting username equal to `${vault.vault-1.smtp.username}` and `${vault.vault-1.smtp.username}`. The Keycloak could extract from this string everything that is needed to to grab this value from a vault, namely:
- Vault id is `vault-1`
- Vault secret id is `smtp.username`

When a realm is loaded, the Keycloak implementation will scan all the fields and if `${vault.*}` is found, use the `VaultServiceProvider` (passing both vault id and vault secret id) to obtain proper values. 

Note, that this implementation doesn't require hardcoded mapping and is very scalable. It can be easily written in a generic way.

Reply:
> I have some doubts how this approach will scale if we add more features to Keycloak. Note, that every time you add something new to Keycloak that requires storing some sort of credentials, you will need to revisit this mapping and add new parameters to it.

No, this is auxiliary class with constants that need not be used but is beneficial. One can still use plain `String` when desired. However, it is better to have compiler check that we use proper context, compiler would not fail for the following:

```java
 sp.obtainSecret(context, "vault-id", "stmp");
```
but would fail for
```java
 sp.obtainSecret(context, "vault-id", Purpose.STMP_SECRET);
```
while it is still possible to write:
```java
 sp.obtainSecret(context, "vault-id", "smtp");
```
but rather error-prone. Why would not this scale?

> A Keycloak Admin could do it in the Email configuration screen by setting username equal to `${vault.vault-1.smtp.username}` [...]

The drawback of the approach with parsed `vault.vault-1.smtp.username` is that `vault-1` allows for no specification of the purpose and uses strictly that vault for `smtp.username` secret. Changing vaults would not be possible with this approaches without changing the entered properties.

Reply:
> Why would not this scale?

@hmlnarik I'm looking at this problem from the development perspective. Let's assume that we would like to introduce a new feature that uses some sort of credentials (that we want to keep confidential, thus we want to use Vault to store them). Implementing such a new feature would require adding a new mapping, e.g. `myfeature => vault-5`. This mapping will be interpreted by some piece of Java code. This means, that every new feature that wants to use Vault, would need to modify that piece of Java code and add its own mapping there.

My proposal (described the above), doesn't require modifying such a mapping. You could use Vault out of the box anywhere in the configuration. You just need to follow some sort of naming convention. So for a `myfeature` example, I would just need to configure credentials in the Admin Console like this: `${vault.vault-5.myfeature.credentials}`.

Just to clear it even more, the whole interaction across Keycloak would look like this (it's a bottom up point of view):

```
MyFeaureModelProvider
     | <-- Here's where we query for ${vault.vault5.myfeature.credentials}
     | <-- and obtain de-vaulted credentials
VaultProvider (has a map of all vaults)
     |
VaultServiceProvider (implements MyVault)
     |
MyFeatureModel (an Entity used by MyFeature)
     |
Hibernate
     |
Database
```

So in this context, whenever I mention it "scales", I mean, "how easy it is to add a new feature with credentials stored in a vault".

Reply:
To sum up an offline discussion - if desired, all of this could be done inside a special vault provider instead of a built-in solution that would mandate a specific property naming.